### DISCUSSION TITLE
Getting helpful error message from `Json` `with_rejection`

### DISCUSSION THREAD

**Discussion by tombowditch** [tokio-rs/axum#1610] [Q&A]
Hi all,

If I give an invalid JSON input in a request body, I get the following error in the response body:
```
Failed to deserialize the JSON body into the target type: accountProvider: unknown variant `STRIPEa`, expected `STRIPE` at line 2 column 30
```

This is a super helpful error as it tells you exactly where you are going wrong.

I however want to return this error as JSON in my API.

I've followed a similar pattern to the with_rejection example here: https://github.com/tokio-rs/axum/blob/main/examples/customize-extractor-error/src/with_rejection.rs and am now up to a point where I have a JsonRejection. The type here of JsonRejection here is a JsonDataError. If I return something (in json) like `e.to_string()`, where `e` is JsonDataError the message is `Failed to deserialize the JSON body into the target type`

If I log out `e` as Debug, I get a `JsonDataError(Error { inner: Error { path: Path { segments: [Map { key: "accountProvider" }] }, original: Error("unknown variant `STRIPEa`, expected `STRIPE`", line: 2, column: 30) } })`

My question is - how do I get the text from the `original` section of that debug output? I can downcast as axum::Error which takes me one step closer, but I am not seeing how to get the actual text I want out of it.

--

My current thoughts is to use a custom extractor ( https://github.com/tokio-rs/axum/blob/main/examples/customize-extractor-error/src/custom_extractor.rs ) instead and do the serde derserialization myself, but I would like to avoid that if possible and let axum do most of the work for me seeing as it is already implemented.

**Comment by davidpdrsn** [comment]
Calling `.to_string()` on a `JsonRejection` should give you that whole error text you see. Then you can wrap that in `Json`.

**Reply by tombowditch** [reply]
Calling `.to_string()` on `JsonRejection` gives me `Failed to deserialize the JSON body into the target type`

I'd like to get the full ```Failed to deserialize the JSON body into the target type: accountProvider: unknown variant `STRIPEa`, expected `STRIPE` at line 2 column 30``` if possible as the other half of the message is useful to see what went wrong.

**Reply by davidpdrsn** [reply]
Ah yeah I see that.

Its because the `IntoResponse` impl and the `Display` text doesn't quite match. I think we should add a method to access to body text. I've filed https://github.com/tokio-rs/axum/issues/1611 about it.

However I would still generally recommend what [jplatte says](https://github.com/tokio-rs/axum/discussions/1610#discussioncomment-4292449).

**Comment by jplatte** [comment]
I would recommend copy-pasting our `Json` type's `FromRequest` implementation and modifying it rather than calling it internally and working off of `JsonRejection`. Then you can have a more structured JSON response.

**Comment by dxps** [comment]
@jplatte or @davidpdrsn Do we have an example for this use case, please :grey_question:  I'll still search for it.
Of course, just copy-pasing that implementation and adapting a bit is not enough since:
```
error[E0210]: type parameter `S` must be used as the type parameter for some local type (e.g., `MyStruct<S>`)
  --> backend/src/handlers/json_rejection.rs:16:9
   |
16 | impl<T, S, B> FromRequest<S, B> for Json<T>
   |         ^ type parameter `S` must be used as the type parameter for some local type
   |
   = note: implementing a foreign trait is only possible if at least one of the types for which it is implemented is local
   = note: only traits defined in the current crate can be implemented for a type parameter
```

Update: 
I believe this is a pretty useful and very reusable concern on many APIs built on top of Axum. So we should have it easily doable.
I guess this [example WithRejection](https://github.com/tokio-rs/axum/pull/1276/files#diff-09342f1dd174e7afeeca86f47890dd3155cd7a0f67e3173aedc49c0e85603699) should work. I'm looking into it ...

**Reply by davidpdrsn** [reply]
https://github.com/tokio-rs/axum/tree/main/examples/customize-extractor-error

**Reply by dxps** [reply]
For now, tried both [custom extractor](https://github.com/tokio-rs/axum/blob/main/examples/customize-extractor-error/src/custom_extractor.rs) and [`WithRejection`](https://github.com/tokio-rs/axum/blob/main/examples/customize-extractor-error/src/with_rejection.rs) examples, and while both does the job of returning a JSON (as content-type and body), unfortunately the root cause of the error (that is captured in the classic setup) remains hidden.

Concretely speaking, if the request payload cannot be deserialized:
- in the classic Axum setup, the (text-based) response is:
```
Failed to deserialize the JSON body into the target type: user: missing field `password` at line 4 column 5
```
- using either of the aforementioned examples, the (JSON-based) response is:
```json
{
  "message": "Failed to deserialize the JSON body into the target type",
  "origin": "custom_extractor",
  "path": "/api/users/login"
}
```

**Reply by davidpdrsn** [reply]
What is your question?

You can access inner errors like this https://docs.rs/axum/0.6.2/axum/extract/index.html#accessing-inner-errors

**Reply by dxps** [reply]
My bad! :blush:  It's all fine now.

I think I should write a small article for others to understand this common concern.

I'm so grateful for getting quick and straight to the point answers from you and for picking up Axum. :pray: 

Cheers! And keep up the good work!

**Reply by hannandesai** [reply]
@dxps you can get the classic error message by calling body_text() function on error like below

```
match payload {
        Ok(Json(payload)) => {
            Ok(payload)
        },
        Err(err) => {
            println!("Payload validate error: {}", err.body_text());
            Err(err.body_text())
        }
    }
```

**Reply by hxk1633** [reply]
@hannandesai Is there a way to do this in a cleaner way rather than doing pattern matching in every handler function for the API routes? Perhaps using middleware function? If so, an example would be really helpful to see how it's done.
